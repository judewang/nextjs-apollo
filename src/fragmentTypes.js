import set from 'lodash/set';
import get from 'lodash/get';
import filter from 'lodash/filter';
import { graphql } from 'graphql';

export default function builder(schema) {
  return graphql(schema, '{ __schema { types { kind name possibleTypes { name } } } }')
    .then(({ data }) => {
      const types = filter(get(data, ['__schema', 'types']), ({ possibleTypes }) => possibleTypes !== null);
      set(data, ['__schema', 'types'], types);
      return JSON.stringify(data);
    });
}

export function execFile(schema) {
  return builder(schema)
    .then(data => data
      .replace(/"([_\w]+)":/gi, (...args) => ` ${args[1]}: `)
      .replace(/"([_\w]+)"/gi, (...args) => `'${args[1]}'`)
      .replace(/([^ ])}/gi, (...args) => `${args[1]} }`)
      .replace(/([^ ])}/gi, (...args) => `${args[1]} }`)
      .replace(/,([^ ])/gi, (...args) => `, ${args[1]}`))
    .then(data => `export default ${data};\n`);
}
