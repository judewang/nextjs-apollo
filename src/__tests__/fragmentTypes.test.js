import {
  GraphQLID, GraphQLInt, GraphQLString, GraphQLSchema, GraphQLInterfaceType, GraphQLObjectType,
} from 'graphql';
import fragmentTypes, { execFile } from '../fragmentTypes';

const Animal = new GraphQLInterfaceType({
  name: 'Animal',
  fields: {
    id: { type: GraphQLID },
    name: { type: GraphQLString },
  },
  resolveType({ type }) {
    return type;
  },
});

const Cat = new GraphQLObjectType({
  name: 'Cat',
  interfaces: [Animal],
  fields: {
    id: { type: GraphQLID },
    name: { type: GraphQLString },
    group: { type: GraphQLInt },
  },
});

const Dog = new GraphQLObjectType({
  name: 'Dog',
  interfaces: [Animal],
  fields: {
    id: { type: GraphQLID },
    name: { type: GraphQLString },
    home: { type: GraphQLString },
  },
});

const Sunflower = new GraphQLObjectType({
  name: 'Sunflower',
  fields: {
    id: { type: GraphQLID },
    createdAt: { type: GraphQLString },
  },
});

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      cat: { type: Cat },
      dog: { type: Dog },
      sunflower: { type: Sunflower },
    },
  }),
});

describe('fragmentTypes', () => {
  it('successfully generated json', async () => {
    expect(await fragmentTypes(schema)).toMatchSnapshot();
  });

  it('execFile', async () => {
    expect(await execFile(schema)).toMatchSnapshot();
  });
});
