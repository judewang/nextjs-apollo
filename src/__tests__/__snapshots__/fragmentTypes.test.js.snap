// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragmentTypes execFile 1`] = `
"export default { __schema: { types: [{ kind: 'INTERFACE', name: 'Animal', possibleTypes: [{ name: 'Cat' }, { name: 'Dog' }] }] } };
"
`;

exports[`fragmentTypes successfully generated json 1`] = `"{\\"__schema\\":{\\"types\\":[{\\"kind\\":\\"INTERFACE\\",\\"name\\":\\"Animal\\",\\"possibleTypes\\":[{\\"name\\":\\"Cat\\"},{\\"name\\":\\"Dog\\"}]}]}}"`;
